compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts

```diff
@@ -0,0 +1,517 @@
+        if (
+          lvalue.identifier.name == null &&
+          value.place.identifier.name !== null &&
+          !usedOutside
+        ) {
+          temporaries.set(lvalue.identifier, value.place.identifier);
+        }
+      }
+    }
+    propertyLoadInfo.set(block.id, propertyLoads);
+  }
+  return {temporaries, properties, propertyLoadInfo};
+}
+
+function collectNodes(
+  fn: HIRFunction,
+  sidemap: TemporariesSidemap,
+): Map<BlockId, BlockInfo> {
+  /**
+   * Due to current limitations of mutable range inference, there are edge cases in
+   * which we infer known-immutable values (e.g. props or hook params) to have a
+   * mutable range and scope.
+   * (see `destructure-array-declaration-to-context-var` fixture)
+   * We track known immutable identifiers to reduce regressions (as PropagateScopeDeps
+   * is being rewritten to HIR).
+   */
+  const knownImmutableIdentifiers = new Set<Identifier>();
+  if (fn.fnType === 'Component' || fn.fnType === 'Hook') {
+    for (const p of fn.params) {
+      if (p.kind === 'Identifier') {
+        knownImmutableIdentifiers.add(p.identifier);
+      }
+    }
+  }
+  const tree = new Tree();
+  const nodes = new Map<BlockId, BlockInfo>();
+  for (const [_, block] of fn.body.blocks) {
+    const propertyLoadInfos = sidemap.propertyLoadInfo.get(block.id);
+    CompilerError.invariant(propertyLoadInfos != null, {
+      reason: '[CollectHoistablePropertyLoads] block info missing',
+      loc: GeneratedSource,
+    });
+
+    const assumedNonNullObjects = new Set<PropertyLoadNode>();
+    for (const {instr, property} of propertyLoadInfos) {
+      const object = instr.value.object.identifier;
+      const propertyNode = tree.getPropertyLoadNode(property);
+      /**
+       * Since this runs *after* buildReactiveScopeTerminals, identifier mutable ranges
+       * are not valid with respect to current instruction id numbering.
+       * We use attached reactive scope ranges as a proxy for mutable range, but this
+       * is an overestimate as (1) scope ranges merge and align to form valid program
+       * blocks and (2) passes like MemoizeFbtAndMacroOperands may assign scopes to
+       * non-mutable identifiers.
+       *
+       * See comment at top of function for why we track known immutable identifiers.
+       */
+      const isMutableAtInstr =
+        object.mutableRange.end > object.mutableRange.start + 1 &&
+        object.scope != null &&
+        inRange(instr, object.scope.range);
+      if (
+        !isMutableAtInstr ||
+        knownImmutableIdentifiers.has(propertyNode.fullPath.identifier)
+      ) {
+        let curr = propertyNode.parent;
+        while (curr != null) {
+          assumedNonNullObjects.add(curr);
+          curr = curr.parent;
+        }
+      }
+    }
+
+    nodes.set(block.id, {
+      block,
+      assumedNonNullObjects,
+    });
+  }
+  return nodes;
+}
+
+function deriveNonNull(fn: HIRFunction, nodes: Map<BlockId, BlockInfo>): void {
```

Comment:
(This follows a non-dominator based approach)

Take this example. Here, we know that it's safe to access `a.b` on line 5 even though no dominator of its basic block accesses a.b.
```js
0: if(...) {
1:   a.b;
2: } else {
3:   a.b;
4: }
5: // It's safe to take `a.b` as a dependency of scope@0 (i.e. hoist the a.b access)
6: scope @0 {
7:   if (cond) {
8:     // access a.b;
9:   }
10:}
```

Currently, we record that `a.b` is safe to read in $bb_i$ if any of the following hold:
- $bb_i$ itself accesses a.b
- a.b is safe to access in every pred of `a.b`
- a.b is safe to access in every successor of `a.b`